apis/Google.Cloud.Storage.V1/Google.Cloud.Storage.V1.IntegrationTests/HmacKeysTest.cs

```diff
@@ -0,0 +1,157 @@
+// limitations under the License.
+
+using Google.Apis.Auth.OAuth2;
+using Google.Apis.Storage.v1.Data;
+using System;
+using System.Linq;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace Google.Cloud.Storage.V1.IntegrationTests
+{
+    [Collection(nameof(StorageFixture))]
+    public class HmacKeysTest
+    {
+        private readonly StorageFixture _fixture;
+
+        public HmacKeysTest(StorageFixture fixture) => _fixture = fixture;
+
+        [Fact]
+        public void Lifecycle()
+        {
+            var client = _fixture.Client;
+            var projectId = _fixture.ProjectId;
+            var serviceAccountEmail = GetServiceAccountEmail();
+            // This is generally a valid service account for a project.
+            string alternativeServiceAccountEmail = $"{projectId}@appspot.gserviceaccount.com";
+
+            // If we find this fails in CI, we'll need to work out an alternative plan.
+            Assert.NotEqual(serviceAccountEmail, alternativeServiceAccountEmail);
+
+            Assert.Empty(client.ListHmacKeys(projectId));
+            var key = client.CreateHmacKey(projectId, serviceAccountEmail);
+            // TODO: When the discovery doc has been updated, this should be an HmacKeyMetadata
+            dynamic metadata = key.Metadata;
+            string accessId = metadata.accessId;
+
+            // We should always get a 40 character secret, which is valid base64.
+            Assert.Equal(40, key.Secret.Length);
+            Convert.FromBase64String(key.Secret);
+
+            // We should now be able to find the key when listing it, either with no filter or filtering by the right email address.
+            var listed = Assert.Single(client.ListHmacKeys(projectId));
+            Assert.Equal(accessId, listed.AccessId);
+            listed = Assert.Single(client.ListHmacKeys(projectId, serviceAccountEmail));
+            Assert.Equal(accessId, listed.AccessId);
+
+            // But not when filtering with the wrong email address.
+            Assert.Empty(client.ListHmacKeys(projectId, alternativeServiceAccountEmail));
+
+            // We should be able to update the key to disable it.
+            // Note: We shouldn't need to create this ourselves...
+            var fixme = new HmacKeyMetadata
+            {
+                ProjectId = projectId,
+                AccessId = accessId,
+                State = "INACTIVE"
+            };
+            var updated = client.UpdateHmacKey(fixme);
+            Assert.Equal(fixme.State, updated.State);
+
+            // We should be able to fetch the key and see that it's been updated
+            var fetched = client.GetHmacKey(projectId, accessId);
+            Assert.Equal(fixme.State, fetched.State);
+
+            // We should be able to delete the key
+            client.DeleteHmacKey(projectId, accessId);
+
+            var deleted = client.GetHmacKey(projectId, accessId);
+            Assert.Equal("DELETED", deleted.State);
```

Comment:
Provide the `State` strings in as `const`s somewhere?

Reply:
Hmm. Can at least do so in the test code. We've got constants in StorageClasses... we could have HmacKeyStates I suppose?